intro/intro_tutorial.py

```diff
@@ -0,0 +1,248 @@
+
+df_train = load_unlabeled_spam_dataset()
+
+# %% [markdown]
+# ## (1) Writing Labeling Functions
+#
+# _Labeling functions (LFs)_ are one of the core operators for building and managing training datasets programmatically in Snorkel.
+# The basic idea is simple: **a labeling function is a function that labels some subset of the training dataset**.
+# That is, in our example here, each labeling function takes as input a comment object, and either outputs a label (`SPAM = 1` or `NOT_SPAM = 0`) or abstains from labeling (`ABSTAIN = -1`):
+
+# %%
+# Define the label mappings for convenience
+ABSTAIN = -1
+NOT_SPAM = 0
+SPAM = 1
+
+# %% [markdown]
+# Labeling functions can be used to represent many heuristic and/or noisy strategies for labeling data, often referred to as [weak supervision](#).
+# The basic idea of labeling functions, and other programmatic operators in Snorkel, is to let users inject domain information into machine learning models in higher level, higher bandwidth ways than manually specifying single-bit labels for individual data points.
+# **The key idea is that labeling functions do not need to be perfectly accurate**, and can in fact even be  correlated with each other, as Snorkel will automatically estimate their accuracies and correlations in a [provably consistent way](https://papers.nips.cc/paper/6523-data-programming-creating-large-training-sets-quickly), and then reweight and combine their output labels, leading to high-quality training labels.
+
+# %% [markdown]
+# In our text data setting here, labeling functions can be keyword matchers:
+
+# %%
+from snorkel.labeling import labeling_function
+
+
+@labeling_function()
+def lf_keyword_my(x):
+    """Many spam comments talk about 'my channel', 'my video', etc."""
+    return SPAM if "my" in x.text.lower() else ABSTAIN
+
+
+# %% [markdown]
+# Use regular expressions:
+
+# %%
+import re
+
+
+@labeling_function()
+def lf_regex_check_out(x):
+    """Spam comments say 'check out my video', 'check it out', etc."""
+    return SPAM if re.search(r"check.*out", x.text, flags=re.I) else ABSTAIN
+
+
+# %% [markdown]
+# Flexibly leverage arbitrary heuristics:
+
+# %%
+@labeling_function()
+def lf_short_comment(x):
+    """Non-spam comments are often short, such as 'cool video!'."""
+    return NOT_SPAM if len(x.text.split()) < 5 else ABSTAIN
+
+
+# %% [markdown]
+# Use third-party models:
+
+# %%
+from textblob import TextBlob
+
+
+@labeling_function()
+def lf_textblob_polarity(x):
+    """
+    We use a third-party sentiment classification model, TextBlob.
+
+    We combine this with the heuristic that non-spam comments are often positive.
+    """
+    sentiment = TextBlob(x.text).sentiment
+    return NOT_SPAM if sentiment.polarity > 0.3 else ABSTAIN
+
+
+# %% [markdown]
+# And much more!
+# For many more types of labeling functions---including over data modalities beyond text---see the other [tutorials](#) and [real-world examples](#).
```

Comment:
mdash instead of triple

Reply:
*Em*-dash...

Reply:
Either way, that felt good, but how do you type one?

Reply:
Alt-"-" correct?

Reply:
lol @vincentschen ????

Reply:
aka vinny aka mdash